### ISSUE TITLE
Consider `requestRequiredInputs` concept

### ISSUE THREAD

**Issue by dglazkov** [google/labs-prototypes#250]
What if we were a bit more lenient and flexible about required inputs? For example, currently in `palm-generateText` node, we currently [throw](https://github.com/google/labs-prototypes/blob/main/seeds/palm-kit/src/nodes/generate-text.ts#L55) when `PALM_KEY` input is not present:

```ts
if (!values.PALM_KEY)
    throw new Error("Text completion requires `PALM_KEY` input");
```

This is a hard error, and recovery from it requires modifying the graph. What if instead, we did something like this:

```ts
if (!values.PALM_KEY) {
   values.PALM_KEY = await context.requestRequiredInputs({ PALM_KEY: palmKeySchema });
```

Now, it is up to the `NodeHandlerContext` to figure out what to do with the input. It can throw and produce the same result as today.

Or it could just ask the user to fill in the key.

Or it could do that plus add a warning saying that this input wasn't wired into the graph -- and maybe even highlight where the mistake might be?

In the last two cases, the board run can still keep going, recovering from a minor stumble.

What's even cooler is that it gives us a way to let `invoke` ask for inputs that might not be wired into it.

For instance, suppose we have a `json-generator` [graph](https://github.com/google/labs-prototypes/blob/main/seeds/breadboard-web/src/boards/json-generator.ts) that takes in ` generator` input that has to be a board with these inputs:

- it needs to have a `text` input 
- it needs to have a `useStreaming` input.

This shape currently matches the [openai-gpt-35-turbo](https://github.com/google/labs-prototypes/blob/main/seeds/breadboard-web/src/boards/openai-gpt-35-turbo.ts) board and hopefully boards for other models that we'll add in the future.

But what if the generator board wants to ask for more parameters? What if this board is actually an omnibus generator where you can choose from various models?

In such cases, the `invoke` will fail, since this input won't be wired into it.

However, if we change `invoke` to use `run` rather than `runOnce` and call `requestRequiredInputs`, we get the behavior we want: the user will be asked for the extra inputs and `invoke` will run as intended.

**Comment by dglazkov** [comment]
Also, because `invoke` does this by default, the `requestRequiredInputs` bubbles upwards in nested `invoke`s and allow boards that are nested N-deep to still ask for user input.

We could probably add a property to `invoke` that stops the bubbling, just in case developers want to avoid unexpected inputs.

**Comment by seefeldb** [comment]
We already have a lot of these pieces laying around, let's see what is missing:

- Lambdas allow binding other inputs to e.g. a generator, so that we don't have to extend the part that only knows about `text` and `useStreaming`. We still need to detect when one is missing, but it gives a hint as to where the input should be added.
- Exceptions are converted to `$error` outputs, so can be handled within the flow. Crucially this also works for `invoke` and it bubbles up. That's how the retry board works: It gets a lambda that accepts a prompt, and it'll keep appending the previous completion (sent as part of $error), the error message and instructions to retry when the invoke goes to the $error wire. This is a different kind of recovery problem, but I'd like to go for something as general as this, i.e. where logic implemented as a board can help recover.
- What is missing is pausing a board / retrying a specific node. The above is a flow _after_ the error. The retry board starts the whole lambda again (which makes sense in that case, because it's built to act on llm generated outputs, so it would have to go back a few steps anyway). Because lambdas already have all other inputs bound, this works. But maybe we need something more granular?

So just by making errors maybe a bit more structured, maybe an explicit "this information is missing" and maybe a facility to add that wire would be interesting? maybe overkill, but i like the principle that boards can "fix" other boards. sets us up well for when boards are themselves llm-made.

---

The other thought that comes to mind:
- I like the idea of asking users for more input from nested flows. And since inputs to boards have that double meaning already, building on that mechanism, as you suggest in your second comment. This doesn't require the complexity from above, it's just a special feature of `input`. Should we just do that?
- For missing wires, I see two case that are very different:
  1) Data is missing that would come from a fresh node that provides just that piece of information, just like the secrets node. In practice, I think many boards would just always omit those wires and the new logic is "data can either come via explicit wires or be requested by name from the environment". That is, we just got environment variables. Maybe that's then what it should be. Also, with the ability to set them just for boards downstream instead of everywhere. This is how you'd configure the default model.
  2) A wire is missing from another node that actually has that piece of information. Here I think we should break and fix the board, possibly automatically as gestured above.

In summary:
- Yes, bubbling up missing inputs would be very interesting
- Sounds like some scoped env variable thing might be useful, and it might look like something a node could request.
- Maybe build out $error for missing wires cases

WDYT?

**Comment by dglazkov** [comment]
Closing in favor of #252.
